### DISCUSSION TITLE
如果能够实现智能解析到最快ip就好了

### DISCUSSION THREAD

**Discussion by wangyaojiu** [XIU2/CloudflareSpeedTest#358] [讨论]

**Comment by XIU2** [comment]
> 抱歉回答晚了，才发现我这个项目的通知设置的有问题，导致你们发 Issues 后我没有收到邮件通知。。。  

DNS 服务程序众多，该软件不可能去针对性的兼容 DNS 程序。  
但是该软件可以用**命令行执行（这样设计就是为了方便配合自动化操作）**。  

所以可以用 Shell 脚本来运行 CloudflareST（记得加上 `-p 0` 这样运行后就直接退出了），然后获得测速后结果，再去 DNS 程序配置中将一些域名解析指向 IP。

**Comment by iamgit66** [comment]
Shell 脚本来运行 ，避免输出结果需要手动回车退出，我用了下面语句，由echo发送一个回车。
 echo "\n" | `./CloudflareST -n 500 -t 4 -tp 443 -dn 20 -dt 10 -f ip.txt -p 20   -o result.csv -dd  >top`

**Comment by XIU2** [comment]
@iamgit66 你把 `-p 20` 改成 `-p 0` 就不会显示结果了，而是直接退出程序（从 result.csv 文件中获取输出结果）。

**Comment by XIU2** [comment]
@iamgit66 另外，你这一大串参数 `-n 500 -t 4 -tp 443 -dn 20 -dt 10 -f ip.txt -p 20 -o result.csv` 全都是默认值，都是可以省去的，你这个命令可以精简一下：
```
./CloudflareST -n 500 -t 4 -tp 443 -dn 20 -dt 10 -f ip.txt -p 20 -o result.csv -dd

./CloudflareST -p 0 -dd
```

**Comment by iamgit66** [comment]
确实，我改改

**Comment by XIU2** [comment]
@iamgit66 我又意外发现了一个 BUG， `-p 0` 时没有直接退出程序，v1.3.3 修复了。  
以前我定的是 `-p -1` 时直接退出程序，后来给改成 `-p 0` 了，结果有处代码我给忘改了。。。

**Comment by iamgit66** [comment]
我必须用-p 20 其实用取前10个IP也足够，前10里面大机率有优质的ip。因为我不使用result.csv，所以把显示输出的ip>>到新的top文件，再测试top里的结果

**Comment by XIU2** [comment]
@iamgit66 好吧，怎么使用都行，其实 result.csv 的作用就是用来让大家进一步筛选处理的。  

其实你使用 `head -n 11 result.csv | tail -n 10` 来读取文件内的前十行结果，不需要完全读取所有文件，这样效果是一样的。  
然后可以用 `awk -F ',' '{print $1}'` 来列分割处理。

**Comment by XIU2** [comment]
@iamgit66 

v1.3.4 支持指定**延迟和速度**条件了，只输出同时满足条件的 IP，这样你就不用去碰运气了 `前10里面大机率有优质的ip` ！

**Comment by XIU2** [comment]
@iamgit66 

v1.3.5 优化了下，现在 Windows 以外的系统**不再需要**按下 回车键 或 Ctrl+C 退出了。  
`这个按键退出主要是为了 Windows 下双击运行程序后，可以看到直接输出结果，Linux 等系统就没必要了。`

**Comment by iamgit66** [comment]
同步更新，获取的IP看似速度有提高！

**Comment by fscarmen** [comment]
请问mac如何用？提示：open ip.txt: no such file or directory。已经把ip.txt放在同一个目录了

**Comment by XIU2** [comment]
@fscarmen  emmm 说实话我没用过 MAC，虽然我编译了 MAC 版本，但并没有在 MAC 下测试过（只测试了 Windows Linux）...

你试过用启动参数 -f "ip.txt 的绝对路径" 运行吗？

**Comment by fscarmen** [comment]
> @fscarmen emmm 说实话我没用过 MAC，虽然我编译了 MAC 版本，但并没有在 MAC 下测试过（只测试了 Windows Linux）...
> 
> 你试过用启动参数 -f "ip.txt 的绝对路径" 运行吗？

可以。正在运行了，不过挺麻烦的，因为CloudflareST和ip.txt前面都要加个绝对路径。

**Comment by XIU2** [comment]
@fscarmen 主要是我没用过 MAC（Windows、Linux 上没有这种问题），所以也不知道为什么相对路径不行（一般是工作目录方面的问题），因此也不知道该如何修复。

**Comment by XIU2** [comment]
@fscarmen 

**“CloudflareST和ip.txt前面都要加个绝对路径”**

你具体是怎么运行的？是在**软件所在目录下**运行的吗？  

在 Windows、Linux 系统中，如果终端没有进入 CloudflareST 所在文件夹，直接用绝对路径运行 CloudflareST 的话，都会提示找不到 ip.txt 文件（因为这时相对路径是基于你当前所在文件夹的）。

**Comment by fscarmen** [comment]
1.最好能做个简单的界面就好了
2.出来的节点172开头的ping最低，但连不上的用不了，下载也都是几乎为0

我把csv文件发你的邮箱呢。

**Comment by fscarmen** [comment]
XIU2 <notifications@github.com> 于2020年12月12日周六 下午11:30写道：

**Comment by XIU2** [comment]
@fscarmen 这是个控制台程序，所以不可能再加个 GUI 的，更别说跨平台 GUI了，那等于重写了。

至于中文乱码我在 Windows、Linux 下测试都是正常的，用的也是 UTF-8 编码。

并不是所有 Cloudflare IP 都可以用于指向域名的，如果下载测速是 0.00 说明该 IP 不支持。  

另外延迟低也不代表就是最好的，但是延迟低出现速度快的几率更高。  
你可以设置个条件 `-sl 5` 这样就过滤掉下载速度低于 5MB/s 的 IP 了，建议同时使用 `-tl 300` 来过滤平均延迟高于 300ms 的 IP，避免一直凑不够 5MB/s 以上 IP 而一直测速下去（毕竟延迟太高的 IP，速度基本上也捉急）。
> `-dn 20` 是数量，默认是 20 个，在满足延迟和速度的条件下凑够 20 个才会停止测速。
